src/node/tendermint_internals.ml

```diff
@@ -0,0 +1,158 @@
+open Crypto
+open Protocol
+open Validators
+
+type value =
+  | Block of Protocol.Block.t
+  | Nil
+[@@deriving yojson]
+
+let string_of_value = function
+  | Nil -> "nil"
+  | Block b ->
+    Printf.sprintf "block %s" (Crypto.BLAKE2B.to_string b.Protocol.Block.hash)
+
+let repr_of_value v = v
+
+let produce_value : (State.t -> value) ref = ref (fun _ -> assert false)
+
+let is_valid state value =
+  match value with
+  | Nil -> false
+  | Block block ->
+    let all_operations_properly_signed _block = true in
+    block.Block.block_height >= state.State.protocol.block_height
+    && all_operations_properly_signed block
+
+let on_block ~nil_default f = function
+  | Nil -> nil_default
+  | Block b -> f b
+
+let block b = Block b
+let nil = Nil
+
+let update_value value round =
+  match value with
+  | Nil -> failwith "Can't update nil value"
+  | Block b -> Block (Protocol.Block.update_round b ~consensus_round:round)
+
+type height = int64 [@@deriving yojson]
+
+type round = int [@@deriving yojson]
+
+type consensus_step =
+  | Proposal
+  | Prevote
+  | Precommit
+
+let string_of_step = function
+  | Proposal -> "Proposal"
+  | Prevote -> "Prevote"
+  | Precommit -> "Precommit"
+
+type sidechain_consensus_op =
+  | ProposalOP  of (height * round * value * round)
```

Comment:
There are two values of the same type. This is begging for them to be confused and then misused. I suggest a record instead of a tuple.